src/machi_lifecycle_mgr.erl

```diff
@@ -0,0 +1,1079 @@
+-compile(export_all).
+-endif. % TEST
+
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-define(SERVER, ?MODULE). 
+
+-record(state, {
+          flus = []   :: [atom()],
+          chains = [] :: list()
+         }).
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+process_pending() ->
+    gen_server:call(?SERVER, {process_pending}, infinity).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+init([]) ->
+    self() ! finish_init,
+    {ok, #state{}}.
+
+handle_call({process_pending}, _From, State) ->
+    {Reply, NewState} = do_process_pending(State),
+    {reply, Reply, NewState};
+handle_call(_Request, _From, State) ->
+    Reply = 'whatwatwha????????????????????',
+    {reply, Reply, State}.
+
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info(finish_init, State) ->
+    {noreply, finish_init(State)};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+finish_init(S) ->
+    %% machi_flu_sup will start all FLUs that have a valid definition
+    %% file.  That supervisor's structure + OTP supervisor behavior
+    %% guarantees that those FLUs should now be running.
+    %% (TODO: Unless they absolutely cannot keep running and the
+    %% supervisor has given up and terminated them.)
+    RunningFLUs = get_local_running_flus(),
+    RunningFLU_Epochs = get_latest_public_epochs(RunningFLUs),
+    RunningFLUs_at_zero = [FLU || {FLU, 0} <- RunningFLU_Epochs],
+    lager:info("Running FLUs: ~p\n", [RunningFLUs]),
+    lager:info("Running FLUs at epoch 0: ~p\n", [RunningFLUs_at_zero]),
+    ChainDefs = get_initial_chains(),
+    perhaps_bootstrap_chains(ChainDefs, RunningFLUs_at_zero, RunningFLUs),
+    S#state{flus=RunningFLUs, chains=ChainDefs}.
+
+get_local_running_flus() ->
+    [Name || {Name,_,_,_} <- supervisor:which_children(machi_flu_sup)].
+
+get_latest_public_epochs(FLUs) ->
+    [begin
+         PS = machi_flu1:make_projection_server_regname(FLU),
+         {ok, {Epoch, _CSum}} = machi_projection_store:get_latest_epochid(
+                                  PS, public),
+         {FLU, Epoch}
+     end || FLU <- FLUs].
+
+get_initial_chains() ->
+    ConfigDir = get_chain_config_dir(),
+    CDefs = [CDef || {_File, CDef} <- machi_flu_sup:load_rc_d_files_from_dir(
+                                        ConfigDir)],
+    sanitize_chain_def_records(CDefs).
+
+sanitize_chain_def_records(Ps) ->
+    {Sane, _} = lists:foldl(fun sanitize_chain_def_rec/2, {[], gb_trees:empty()}, Ps),
+    Sane.
+
+sanitize_chain_def_rec(Whole, {Acc, D}) ->
+    try
+        #chain_def_v1{name=Name,
+                      mode=Mode,
+                      full=Full,
+                      witnesses=Witnesses} = Whole,
+        {true, 10} = {is_atom(Name), 10},
```

Comment:
Is the use of the integers here (e.g., `10`, `20`, ...) for debugging purposes or some other purpose?  It would be nice if you would add a comment explaining what they're for.